### PR TITLE
INT-3541: The XMPP Extension support on outbound

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/INT-3541

### CODE REVIEW

**Comment by artembilan** [comment]
Hello, @puel !

We would be glad if you'd take a look here and provide your feedback for this new `extension-provider` feature on the Outbound Channel Adapter.

Thank you anyway!

File: spring-integration-xmpp/src/main/java/org/springframework/integration/xmpp/outbound/ChatMessageSendingMessageHandler.java [hunk 1]
```diff
@@ -40,8 +47,11 @@
  */
 public class ChatMessageSendingMessageHandler extends AbstractXmppConnectionAwareMessageHandler {
 
+	private static final Pattern xmlPattern = Pattern.compile("<(\\S.*?)>(.*?)</\\1>");
```
Comment (garyrussell) [inline spring-integration-xmpp/src/main/java/org/springframework/integration/xmpp/outbound/ChatMessageSendingMessageHandler.java]: `[^>]*>[^<]*<` is generally more efficient than `.*?>.*?` - see [here](http://blog.stevenlevithan.com/archives/greedy-lazy-performance).

You also don't need to capture the value, so I would recommend

`"<(\\S[^>]*)>[^<]*</\\1>"`

Perhaps we should also allow leading and trailing spaces (`\\s*`).

Comment (garyrussell) [inline spring-integration-xmpp/src/main/java/org/springframework/integration/xmpp/outbound/ChatMessageSendingMessageHandler.java]: Is it possible the tag might have attributes? If so, this won't work, we'd need something like

`<(\\S[^>\\s]*)[^>]*>`

File: spring-integration-xmpp/src/main/java/org/springframework/integration/xmpp/outbound/ChatMessageSendingMessageHandler.java [hunk 2]
```diff
@@ -64,26 +85,47 @@ public String getComponentType() {
...
-				this.headerMapper.fromHeadersToRequest(message.getHeaders(), xmppMessage);
+
+			if (payload instanceof ExtensionElement) {
+				xmppMessage.addExtension((ExtensionElement) payload);
+			}
+			else if (payload instanceof String) {
+				if (this.extensionProvider != null) {
+					String data = (String) payload;
+					if (!xmlPattern.matcher(data).matches()) {
+						data = "<root>" + data + "</root>";
```
Comment (garyrussell) [inline spring-integration-xmpp/src/main/java/org/springframework/integration/xmpp/outbound/ChatMessageSendingMessageHandler.java]: We should add a comment that this is just an arbitrary tag that is removed by the extension parser.

**Comment by garyrussell** [comment]
That's all

**Comment by artembilan** [comment]
Pushed.
Thank you, Gary, for sharing your wisdom.
I provided your solution with some polishing.
Please, see the commit changes.

**Comment by garyrussell** [comment]
Rebased, squashed, merged as b363bdb15eee00dd4781c8b48f7973f5c594c7f4

**Pull request closed without merging**